### ISSUE TITLE
Detection 1: Changing LAC/CID

### ISSUE THREAD

**Issue by He3556** [SecUpwN/Android-IMSI-Catcher-Detector#91]
The LAC (Location Area Code) describes a set of Cell Towers (with different IDs) like below:

![step_01](https://cloud.githubusercontent.com/assets/2507905/3635662/df644cfc-0f82-11e4-85c3-bfcfd0a45195.jpg)

**Step 1** of the IMSI-Catcher *(picture 1)*: Masquerade like a real BTS and send with more power than the original, so a cell phone would connect to it. But if the connection is established it is still using the TIMSI (Temporary IMSI).

![step_02](https://cloud.githubusercontent.com/assets/2507905/3635663/e485f8e8-0f82-11e4-9365-76943a8d72d7.jpg)

**Step 2** of the IMSI-Catcher *(picture 2)*: One possibility to get the IMSI:
* Change the LAC of the BTS *(picture 2)*, so that the Location Update Procedure is initiated. 

![unbenannt](https://cloud.githubusercontent.com/assets/2507905/3635643/988007a0-0f81-11e4-872a-7cb5728a6e7d.png)

If the MSC (Controller of a group of BTS) is also changing with this location update, then the phone would have to send the IMSI. Or if the location update fails it will also send the IMSI.
See Figure 4.1.1.1 [http://www.qtc.jp/3GPP/Specs/23012-520.pdf]

However, the Catcher-Catcher Project gives a yellow or a even a red flag if the LAC is changing, so we really should implement this.  It is also quite simple to do, because we have the values LAC and CellID. If the CellID changes the LAC over the time – we can show a yellow flag – if it changes more than once we show a red flag. This happens while the IMSI-Catcher is catching IMSI’s – not when a call is established. So you don’t have to be the victim to detect a fake BTS.

**Comment by He3556** [comment]
We shouldn't put 2 detections in one Issue again. This was "Changing LAC" and "Wrong CID" - now that we are finished with it, i will keep it like it is. 

But there is still a bug giving wrong alerts (changing LAC) under special conditions. 
(more logcat added in the last release - hope that helps to find the bug) 
We have an extra issue for that, so we could close this one here i guess.

**Comment by SecUpwN** [comment]
>We have an extra issue for that, so we could close this one here i guess.

Which one? Please reference the correct Issue and close this if really done.

**Comment by He3556** [comment]
There is one think ToDo before we can close this Issue.
Missing return [here](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/blob/master/app/src/main/java/com/SecUpwN/AIMSICD/adapters/AIMSICDDbAdapter.java#L187-L188.) to update the Notification somewhere [here](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/blob/master/app/src/main/java/com/SecUpwN/AIMSICD/service/CellTracker.java#L641-L692)
If somebody can give me a hint, how i can do it. Would save me some time :)

**Comment by d-mariano** [comment]
I would like to give this a shot.  What would be a suitable notification?  "Cell ID does not exist in openCellId database." ?  Or should I use the log string?

**Comment by He3556** [comment]
Thanks for picking up this one. Yes "Cell ID does not exist in openCellId database" would be great.

**Comment by He3556** [comment]
thank's to d-mariano - we got this Issue finished :) 
We need to think about a solution for the notifications. If there are two detections at the same time, the first one will be overlapped by the second. Maybe we show an array of notifications in a loop? 
Just mentioned it - so we don't forget about it.

**Comment by E3V3A** [comment]
I just posted detailed instructions [here](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/173#issuecomment-70435310) in the "App Test Protocol" #173 issue, on how to test changing LAC.

**Comment by SecUpwN** [comment]
>thank's to d-mariano - we got this Issue finished :)

Thanks to @d-mariano for his funky pull request! @He3556, if this is **really** solved now, please close.

**Comment by He3556** [comment]
we can still enhance this but actually we are finished at that point. So i just close it :)

**Comment by E3V3A** [comment]
@d-mariano 
Something is not right. Did you test it before making PR? Your code is causing the following problems:
 1. When app start for the first time, it immediately goes to YELLOW detection. Not good, but could be the right behavior, once the DB and BTS pin selection/coloring scheme is implemented.  
 2. That annoying not possible to close/quit bug has returned... Not sure if it is only on my device. 

@SecUpwN @He3556 @Ueland  Please test!

**Comment by He3556** [comment]
There is everything ok with the code and i am absolutely sure, that @d-mariano tested it before ;) 
I just added a reminder in another issue about the problem.

REMINDER for the installation & setup wizard (for the first start of the app) 
We need to wait until the db is downloaded before we check the CID against the data of OpenCellID. Otherwise we get a wrong alert just right after the installation.

**Comment by E3V3A** [comment]
Ok, great. Which issue?

**Comment by He3556** [comment]
here: https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/181

**Comment by SecUpwN** [comment]
@E3V3A, the new Issue for the immediate yellow flag is #290. And just to tell you: We already discussed it in the Testroom before you mentioned it here. So is this closed now, or how often do we need to re-open this?

**Comment by E3V3A** [comment]
Should we rename this to "...Changing LAC on same CID"?
(Because CID detection is yet different.)

**Comment by He3556** [comment]
yes thats why i said - in the future we don't throw detections together, in one issue :) if you want pls rename it.

**Comment by E3V3A** [comment]
Great! (The problem was that I originally though of them as connected.)

**Comment by SecUpwN** [comment]
>Great! (The problem was that I originally though of them as connected.)

@E3V3A, I can't follow, please clarify. We had multiple Issues in here and this one is solved?

**Comment by Tonat** [comment]
Hey there, I experience the following:
Without any change of location, seemingly random alerts for "Changing LAC" events (like in https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/439) happen on v0.1.37 on my non-LTE [sic] i9300 with CM 12.1 (without the OCID DBs downloaded due to https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/691). See the log for one of the events posted here: https://defuse.ca/b/jakoA5RPVujZdQ4JOMGEjr. 
I'm not sure whether this may be related to https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/606 as only the time stamp, CID and LAC are showing values in the database viewer while PSC is shown as invalid, DF_id as 1. Lat, Lon and Accu however are all shown as 0. This is the case with all "Changing LAC" events (although the CID sometimes is different). 
   I experienced these events once or twice a day, with not having moved my phone for a while in all cases. 

NB: Note that after updating on v0.1.38 and downloading the OCID DB, this stopped happening (at least it didn't in the past three days). So it may be due to this DB missing.

**Comment by banjaxbanjo** [comment]
@Tonat there was a change in lac as seen your log

01-06 20:00:09.397  10221-10221/com.SecUpwN.AIMSICD I/a﹕ ALERT: Changing LAC on CID: xxxxx3001 LAC(API): abc48 LAC(DBi): abc58

It's  possible that both these LAC locations have the same cell id but if there is no LAC abc58 in your area than I would class this as suspicious.

If this LAC suddenly disappears a few hours later that's even more suspicious.

Try open cell id to see if this lac and cid exist for your area or another website that shows lac and cid locations.

**Comment by He3556** [comment]
If you have this alerts on different CIDs it is the same old problem. Some devices read wrong LAC values from time to time. 
Btw. a CID only changes its LAC when the network gets reconfigured - or when a fake BTS tries to trigger a location update, where your IMSI will be send. 
But it is only 1 fake BTS (CID) with a changing LAC - thats why we should https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/issues/628#issuecomment-156695686

**Comment by limikri** [comment]
> The LAC (Location Area Code) describes a set of Cell Towers (with different IDs) like below:
> 
> ![step_01](https://cloud.githubusercontent.com/assets/2507905/3635662/df644cfc-0f82-11e4-85c3-bfcfd0a45195.jpg)
> 
> **Step 1** of the IMSI-Catcher _(picture 1)_: Masquerade like a real BTS and send with more power than the original, so a cell phone would connect to it. But if the connection is established it is still using the TIMSI (Temporary IMSI).
> 
> ![step_02](https://cloud.githubusercontent.com/assets/2507905/3635663/e485f8e8-0f82-11e4-9365-76943a8d72d7.jpg)
> 
> **Step 2** of the IMSI-Catcher _(picture 2)_: One possibility to get the IMSI:
> - Change the LAC of the BTS _(picture 2)_, so that the Location Update Procedure is initiated. 
> 
> ![unbenannt](https://cloud.githubusercontent.com/assets/2507905/3635643/988007a0-0f81-11e4-872a-7cb5728a6e7d.png)
> 
> If the MSC (Controller of a group of BTS) is also changing with this location update, then the phone would have to send the IMSI. Or if the location update fails it will also send the IMSI.
> See Figure 4.1.1.1 [http://www.qtc.jp/3GPP/Specs/23012-520.pdf]
> 
> However, the Catcher-Catcher Project gives a yellow or a even a red flag if the LAC is changing, so we really should implement this.  It is also quite simple to do, because we have the values LAC and CellID. If the CellID changes the LAC over the time – we can show a yellow flag – if it changes more than once we show a red flag. This happens while the IMSI-Catcher is catching IMSI’s – not when a call is established. So you don’t have to be the victim to detect a fake BTS.  
> ## <bountysource-plugin>
> 
> Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/3203434-changing-lac-on-same-cid?utm_campaign=plugin&utm_content=tracker%2F565750&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F565750&utm_medium=issues&utm_source=github).
> </bountysource-plugin>
>
